File: sdks/go/executions.go
package voidcontrol

import "time"

type ExecutionsClient struct {
	client *Client
}

func (client *ExecutionsClient) Get(executionID string) (*ExecutionDetail, error) {
	var response ExecutionDetail
	if err := client.client.getJSON("/v1/executions/"+executionID, &response); err != nil {
		return nil, err
	}
	return &response, nil
}

func (client *ExecutionsClient) Wait(executionID string) (*ExecutionDetail, error) {
	for {
		detail, err := client.Get(executionID)
		if err != nil {
			return nil, err
		}
		if detail.Execution.Status == "Completed" || detail.Execution.Status == "Failed" || detail.Execution.Status == "Canceled" {
			return detail, nil
		}
		time.Sleep(10 * time.Millisecond)
	}
}

File: sdks/node/src/executions.js
import { toExecutionDetail } from "./models.js";

const TERMINAL_STATUSES = new Set(["Completed", "Failed", "Canceled"]);

export class ExecutionsClient {
  constructor(client) {
    this._client = client;
  }

  async get(executionId) {
    const payload = await this._client.getJson(`/v1/executions/${executionId}`);
    return toExecutionDetail(payload);
  }

  async wait(executionId, { pollIntervalMs = 1000 } = {}) {
    while (true) {
      const detail = await this.get(executionId);
      if (TERMINAL_STATUSES.has(detail.execution.status)) {
        return detail;
      }
      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }
  }
}
